content/blog/2018-08-28-birds-ocr.md

```diff
@@ -0,0 +1,411 @@
+particular, I wanted to try extract names of
+[`magick`](https://github.com/ropensci/magick) for image manipulation,
+[`tesseract`](https://github.com/ropensci/tesseract) for optical
+character recognition,
+[`cld2`](https://github.com/ropensci/cld2)/[`cld3`](https://github.com/ropensci/cld3)
+for language detection… Quite the armory! We’ll also sprinkle some
+taxonomy magic by Scott Chamberlain, one of rOpenSci’s co-founders, to
+resolve the names extracted.
+
+### OCR-bird naming workflow, piece by piece
+
+In this section, we explain the different elements of our R workflow:
+preparing images, extracting text, resolving taxonomic names.
+
+#### Image preparation
+
+I saved the pictures locally in a “birds” folder. Yes, I click-buttoned
+instead of using [the Flickr API](https://www.flickr.com/services/api/)
+for which e.g. Jim Hester wrote a [minimal R
+wrapper](https://github.com/jimhester/flickrr)… I don’t do *everything*
+with R scripts (yet).
+
+``` r
+library("magrittr")
+filenames <- fs::dir_ls("birds")
+```
+
+Each image shows a bird and its name. Images are either landscape- or
+portrait- oriented, but for the sake of simplicity, I’ll act as if they
+were all portrait- oriented. *What I’d do if I weren’t after simplicity
+but still as lazy would be to duplicate images rotated
+(`magick::image_rotate`) in all possible directions, applying the
+workflow, hoping to get good data from one of the rotated versions.*
+
+``` r
+magick::image_read(filenames[1])
+```
+
+<img src="/img/blog-images/2018-08-28-birds-ocr/unnamed-chunk-2-1.png" width="1376" />
+
+From that image, I wanted to extract the name indicated below the bird.
+To maximize the efficiency of Optical Character Recognition, we shall
+first prepare the image, since the [accuracy of OCR depends on the
+quality of the input
+image](https://cran.r-project.org/web/packages/tesseract/vignettes/intro.html#preprocessing_with_magick)
+which we can influence a bit. This part could be tweaked even more, and
+in real life examples it’ll be worth spending time trying different
+`magick` functions and parameter values. Since I have in mind the case
+of a bunch of images to be batch-processed, nothing is done by hand.
+
+``` r
+crop_bird <- function(filename){
+  image <- magick::image_read(filename)
+  
+  height <- magick::image_info(image)$height
+  
+  # crop the top of the image
+  image <- magick::image_crop(image, 
+                     paste0("+0+",round(0.75*height))) %>%
+    # convert the image to black and white
+    magick::image_convert(type = "grayscale") %>%
+    # increase brightness
+    magick::image_modulate(brightness = 120) %>%
+    magick::image_enhance() %>%
+    magick::image_median() %>%
+    magick::image_contrast() 
+  
+  # we'll need the filename later
+  attr(image, "filename") <- filename
+  
+  return(image)
+}
+
+crop_bird(filenames[1])
+```
+
+<img src="/img/blog-images/2018-08-28-birds-ocr/unnamed-chunk-3-1.png" width="1376" />
+
+It does look cleaner now!
+
```

Comment:
need line break before heading (I know these are trouble)

Reply:
whyyy are they so much trouble?! I didn't succeed in getting them right :sob:

Reply:
I tried `<br>` and before that I had tried comparing them to working headers, no luck.